src/TriangleNodesProvisioner.cpp

```diff
@@ -0,0 +1,199 @@
+        int Np = r.length(0);
+        for( index_type i=0; i < Np; i++) {
+            if(s(i) != 1.0) 
+                a(i) = 2.0*(1.0+r(i))/(1.0-s(i)) - 1.0;
+            else
+                a(i) = -1.0;
+        }
+        b = s;
+    }
+
+    void TriangleNodesProvisioner::xyTors(const real_vector_type & x, const real_vector_type & y, real_vector_type & r, real_vector_type & s) const {
+        const index_type Np = x.length(0);
+        real_vector_type L1(Np), L2(Np), L3(Np);
+        L1 = (sqrt(3.0)*y+1.0)/3.0;
+        L2 = (-3.0*x - sqrt(3.0)*y + 2.0)/6.0;
+        L3 = ( 3.0*x - sqrt(3.0)*y + 2.0)/6.0;
+
+        r =-L2 + L3 - L1;
+        s =-L2 - L3 + L1;
+    }
+
+
+    void TriangleNodesProvisioner::computeVandermondeMatrix(const int N, const real_vector_type & r, const real_vector_type & s, const real_matrix_type & V) const {
+        const index_type Nr = r.length(0);
+
+        real_vector_type a(Nr), b(Nr);
+        rsToab(r, s, a, b);
+
+        // build the Vandermonde matrix
+        index_type count = 0;
+        for (index_type i=0; i < N + 1; ++i) {
+            for (index_type j=0; j < N - i + 1; ++j) {
+                real_vector_type p(Nr);
+                evaluateSimplexPolynomial(a,b,i,j,p);
+                V(Range::all(),count) = p;
+                ++count;
+            }
+        }
+    }
+
+    void TriangleNodesProvisioner::computeEquilateralNodes(real_vector_type & x, real_vector_type & y) const {
+        real_vector_type alphaOptimal(15);
+
+        alphaOptimal = 0.0000,0.0000,1.4152,0.1001,0.2751,0.9800,1.0999,
+                       1.2832,1.3648,1.4773,1.4959,1.5743, 1.5770,1.6223,1.6258;
+
+        double alpha = 2.0/3.0;
+        if (NOrder < 16) {
+            alpha = alphaOptimal(NOrder-1);
+        }
+
+        const int Np = (NOrder+1)*(NOrder+2)/2;
+
+        // Create equidistributed nodes on equilateral triangle.
+        real_vector_type L1(Np), L2(Np), L3(Np);
+
+        index_type count = 0;
+
+        for (index_type n=1; n <= NOrder+1; ++n) {
+            for (index_type m=1; m <= NOrder+2-n; ++m) {
+                L1(count) = (n-1.0)/NOrder;
+                L3(count) = (m-1.0)/NOrder;
+                ++count;
+            }
+        }
+        L2 = 1.0 - L1 - L3;
+
+        x = -L2+L3; 
+        y = (-L2-L3+2*L1)/sqrt(3.0);
+
+        // Compute blending function at each node for each edge.
+        real_vector_type blend1(Np), blend2(Np), blend3(Np);
+        blend1 = 4*L2*L3; 
+        blend2 = 4*L1*L3; 
+        blend3 = 4*L1*L2;
+
+        // Amount of warp for each node, for each edge.
+        real_vector_type warpf1(Np), warpf2(Np), warpf3(Np), temp(Np);
+
+        temp = L3 - L2; computeWarpFactor(temp, warpf1);
+        temp = L1 - L3; computeWarpFactor(temp, warpf2);
+        temp = L2 - L1; computeWarpFactor(temp, warpf3);
+
```

Comment:
Why do you need temp here? Just write:
`computeWarpFactor(L3 - L1, warpf1);
computeWarpFactor(L1 - L3, warpf2);
computeWarpFactor(L2  -L1, warpf3);`

Reply:
Will try this, though I was/am under the impression that blitz doesn't accept expressions where it expects arrays.

Reply:
yep, does not compile.

Reply:
My mistake, I forgot about blitz's idiosyncrasies.